### ISSUE TITLE
Question

### ISSUE THREAD

**Issue by Inqus** [navjindervirdee/neural-networks#2]
Thanks for your code, it really helped to parse in the LSTM's, you can’t tell me how to change the code for the regression task, I wanted to change the code for the time series forecast, I changed the function softmax to relu, but it takes the current number as a prediction

**Comment by navjindervirdee** [comment]
I implemented this code in order to learn how the LSTM network works, so this code is specifically designed for classification tasks. 
For regression, 

1. You would need to change the output_cell function. It returns an output of 27 (alphabets + '.') neurons to generate a single value.

2. You need to change the loss function to mean squared error or some other as per your dataset and application.

3. You need to change the output cell derivatives or errors function according to your loss function.

**Comment by Inqus** [comment]
Thanks for reply. Yes, I will change the loss function.
But there is a question about network training (backprop). 
In your example, you are calculating an error (output_cache) after each forecast, but I need to get a forecast after the network has passed all the steps. 
In training, can I calculate error only on the last forecast?

(i should have one activation error at all time steps, but cell error should change gradually, 
and in that case should i use activation_cache (of last timestep) to calculate single lstm cell derivatives )


example
```
 X         Y
[10 20 30] 40
[20 30 40] 50
[30 40 50] 60
[40 50 60] 70
[50 60 70] 80
```
2)
And in your example, 100 input units, but ur input array looks like (max_lenght(timesteps),  batch_size(samples in one batch), vocab_size(27))
why count of input units not equal vocab_size (100 != 27)

**Comment by navjindervirdee** [comment]
In this code the network architecture in one to one i.e for each timestep, one out is produced, and you would need to change it to many to one, i.e one output after all the timesteps. You need to change both the forward propagation and backpropagation.

In the backpropagation process, for each time step error is backpropagated for each previous unfold. You need to change it, such that only error at the last time step is backpropagated to all previous timesteps.

Each char in this code is first represented as a list of 100 neurons called the character embedding which goes as input to the network. Only one char (represented by 100 units) goes as input in each time step and one output of 27 units is produced.

**Comment by Inqus** [comment]
Thanks again for your answer, but there is still a question, I hope you are not fed up with your stupid questions.
I wanted to clarify this function calculates the softmax derivative(output cell)?

```
(old code - your version)
def calculate_output_cell_derivatives(output_error_cache,activation_cache,parameters):
    #to store the sum of derivatives from each time step
    dhow = np.zeros(parameters['how'].shape)
    
     batch_size = activation_cache['a1'].shape[0]
    
    #loop through the time steps 
    for i in range(1,len(output_error_cache)+1):
        #get output error
        output_error = output_error_cache['eo' + str(i)]
        
        #get input activation
        activation = activation_cache['a'+str(i)]
        
        #cal derivative and summing up!
        dhow += np.matmul(activation.T,output_error)/batch_size
        
    return dhow

```

I changed the model's training, and did a couple of tests, with charges of( n_steps,n_features):
1) [simple (1,1)](https://imgur.com/f2eXLap)
2) [window method (1,3)](https://imgur.com/dwt7Y3I)
3) [with time steps (3,1)](https://imgur.com/elVJXSg)
the first two tests showed not bad result, but the last had a terrible result
I got bad results on the 3rd test because the training didn’t do right?

upd. I greatly increased the number of neurons (4 => 50) in the hidden layer and the result with time steps became better, but the error is still quite large (train ~60 rmse, test ~100 rmse) 
it looks like I got stuck somewhere with the training

This is part of the backprop code, each time the activation error is calculated from the error obtained on the output layer
```
    #calculate output cell errors
    def calculate_output_cell_error(self, target):

        self.activation_error_cache = dict()

        #calculate the output_error (ot - predicted) self. ot - output after all time steps
        self.error_output = self.ot - target[:,np.newaxis]

        #calculate the activation error  ea
        # self.error_activation = np.matmul(self.error_output, self.how.T)
        for i in range(1,len(self.activation_cache)):
        
            #calculate the activation error for time step 't'
            error_activation = np.matmul(self.error_output, self.how.T)
        
            #store the output and activation error in dict
            self.activation_error_cache['ea'+str(i)] = error_activation
```

in the case of regression, the calculation of the derivative does not change, right? or am I supposed to get the derivative of the activation function (relu)
```
    #calculate output cell derivatives
    def calculate_output_cell_derivatives(self):
        #to store the sum of derivatives from each time step
        dhow = np.zeros(self.how.shape)
        
        #loop through the time steps 
        for i in range(1,len(self.activation_cache)):
            # get output error (here only error from last output - self.error_output)
            
            # get input activation
            activation = self.activation_cache['a'+str(i)]

            # cal derivative and summing up!
            dhow += np.matmul(activation.T, self.error_output)
            
        return dhow
```
